machine-learning/hyperparam-opt.ipynb

```diff
@@ -0,0 +1,510 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Hyperparameter optimization with Dask\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Every machine learning model has some values that are specified before training begins. These values help adapt the model to the data but must be given before any training data is seen. For example, this might be `penalty` or `C` in Scikit-learn's [LogisiticRegression]. These values that come before any training data and are called \"hyperparameters\". Typical usage looks something like:\n",
+    "\n",
+    "``` python\n",
+    "from sklearn.linear_model import LogisiticRegression\n",
+    "from sklearn.datasets import make_classification\n",
+    "\n",
+    "X, y = make_classification()\n",
+    "est = LogisiticRegression(C=10, penalty=\"l2\")\n",
+    "est.fit(X, y)\n",
+    "```\n",
+    "\n",
+    "These hyperparameters influence the quality of the prediction. If `C` is too small above, the output of the estimator will be too regularized to produce meaningful output.\n",
+    "\n",
+    "Determining the values of these hyperparameters is difficult. In fact, Scikit-learn has an entire documentation page on finding the best values: https://scikit-learn.org/stable/modules/grid_search.html\n",
+    "\n",
+    "[LogisiticRegression]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Dask enables some new techniques and opportunities for hyperparameter optimization. These opportunities involve stopping training early to limit computation (naturally, this requires a `partial_fit` or `warm_start` implementation in Scikit-learn parlance)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**This example will walk through**\n",
+    "\n",
+    "* how to determine the input parameters\n",
+    "* how to use the a particular algorithm implemented in `HyperbandSearchCV`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup Dask"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from distributed import Client\n",
+    "client = Client(\"localhost:8786\")\n",
```

Comment:
This notebook should be runnable on its own without external setup.  We usually create a Client with a specific number of workers and threads-per-worker fully specified with a setup that we know works well.  It's also worth noting that these are likely to run on Binder pods that are pretty low-power.  Typically we pretend that we have something like 4 cores and a couple gigabytes of memory.  

You might want to look at some of the other examples to see how we tend to manage this.

Reply:
Whoops, thanks.